pkg/agent/openflow/pipeline.go

```diff
@@ -2704,3 +2655,233 @@ func (sl conjunctiveActionsInOrder) Less(i, j int) bool {
+	// TODO: load ToUplinkRegMark for packets from Antrea IPAM Pod.
+	return L3ForwardingTable.ofTable.BuildFlow(priorityMiss).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		Action().LoadRegMark(ToGatewayRegMark).
+		Action().GotoStage(stageSwitching).
+		Done()
+
+}
+
+// hostBridgeLocalFlows generates the flows to match the packets forwarded between bridge local port and uplink port.
+func (f *featurePodConnectivity) hostBridgeLocalFlows() []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	return []binding.Flow{
+		// This generates the flow to forward the packets from uplink port to bridge local port.
+		ClassifierTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
+			MatchInPort(config.UplinkOFPort).
+			Action().Output(config.BridgeOFPort).
+			Done(),
+		// This generates the flow to forward the packets from bridge local port to uplink port.
+		ClassifierTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
+			MatchInPort(config.BridgeOFPort).
+			Action().Output(config.UplinkOFPort).
+			Done(),
+	}
+}
+
+// preRoutingClassifierFlows generates the flow to classify packets in stagePreRouting.
+func (f *featureService) preRoutingClassifierFlows() []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var flows []binding.Flow
+
+	targetTables := []uint8{SessionAffinityTable.ofTable.GetID(), ServiceLBTable.ofTable.GetID()}
+	if f.proxyAll {
+		targetTables = append([]uint8{NodePortMarkTable.ofTable.GetID()}, targetTables...)
+	}
+	for _, ipProtocol := range f.ipProtocols {
+		flows = append(flows,
+			// This generates the default flow to match the first packet of Service connection.
+			PreRoutingClassifierTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchProtocol(ipProtocol).
+				Action().ResubmitToTables(targetTables...).
+				Done(),
+		)
+	}
+
+	return flows
+}
+
+// l3FwdFlowsToExternal generates the flows to forward the packets of Service connection to external network.
+func (f *featureService) l3FwdFlowsToExternal() []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var flows []binding.Flow
+	if f.connectUplinkToBridge {
+		flows = append(flows,
+			// When AntreaIPAM is enabled, this generates the flow to match the packets sourced from per-Node IPAM Pod and
+			// destined for external network.
+			L3ForwardingTable.ofTable.BuildFlow(priorityLow).
+				Cookie(cookieID).
+				MatchRegMark(RewriteMACRegMark).
+				MatchRegMark(NotAntreaFlexibleIPAMRegMark).
+				MatchCTMark(ServiceCTMark).
+				Action().SetDstMAC(f.gatewayMAC).
+				Action().LoadRegMark(ToGatewayRegMark).
+				Action().GotoTable(L3DecTTLTable.GetID()).
+				Done(),
+			// When AntreaIPAM is enabled, this generates the flow to match the packets sourced from Antrea IPAM Pod and
+			// destined for external network.
+			L3ForwardingTable.ofTable.BuildFlow(priorityLow).
+				Cookie(cookieID).
+				MatchRegMark(RewriteMACRegMark).
```

Comment:
Could you explain why it matches RewriteMACRegMark but doesn't actually rewrite MAC?

Reply:
RewriteMACRegMark can be removed here.